src/data/nat/enat.lean

```diff
@@ -165,6 +172,74 @@ instance : canonically_ordered_monoid enat :=
   ..enat.semilattice_sup_bot,
   ..enat.ordered_comm_monoid }
 
+protected lemma add_lt_add_right {x y z : enat} (h : x < y) (hz : z ≠ ⊤) : x + z < y + z :=
+begin
+  rcases ne_top_iff.mp (ne_top_of_lt h) with ⟨m, rfl⟩,
+  rcases ne_top_iff.mp hz with ⟨k, rfl⟩,
+  induction y using enat.cases_on with n,
+  { rw [top_add], apply_mod_cast coe_lt_top },
+  norm_cast at h, apply_mod_cast add_lt_add_right h
+end
+
+protected lemma add_lt_add_iff_right {x y z : enat} (hz : z ≠ ⊤) : x + z < y + z ↔ x < y :=
+⟨lt_of_add_lt_add_right', λ h, enat.add_lt_add_right h hz⟩
+
+protected lemma add_lt_add_iff_left {x y z : enat} (hz : z ≠ ⊤) : z + x < z + y ↔ x < y :=
+by simpa using enat.add_lt_add_iff_right hz
+
+protected lemma lt_add_iff_pos_right {x y : enat} (hx : x ≠ ⊤) : x < x + y ↔ 0 < y :=
+by { conv_rhs { rw [← enat.add_lt_add_iff_left hx] }, rw [add_zero] }
+
+lemma lt_add_one {x : enat} (hx : x ≠ ⊤) : x < x + 1 :=
+by { rw [enat.lt_add_iff_pos_right hx], norm_cast, norm_num }
+
+lemma le_of_lt_add_one {x y : enat} (h : x < y + 1) : x ≤ y :=
+begin
+  induction y using enat.cases_on with n, apply lattice.le_top,
+  rcases ne_top_iff.mp (ne_top_of_lt h) with ⟨m, rfl⟩,
+  apply_mod_cast nat.le_of_lt_succ, apply_mod_cast h
+end
+
+lemma add_one_le_of_lt {x y : enat} (h : x < y) : x + 1 ≤ y :=
+begin
+  induction y using enat.cases_on with n, apply lattice.le_top,
+  rcases ne_top_iff.mp (ne_top_of_lt h) with ⟨m, rfl⟩,
+  apply_mod_cast nat.succ_le_of_lt, apply_mod_cast h
+end
+
+lemma add_one_le_iff_lt {x y : enat} (hx : x ≠ ⊤) : x + 1 ≤ y ↔ x < y :=
+begin
+  split, swap, exact add_one_le_of_lt,
+  intro h, rcases ne_top_iff.mp hx with ⟨m, rfl⟩,
+  induction y using enat.cases_on with n, apply coe_lt_top,
+  apply_mod_cast nat.lt_of_succ_le, apply_mod_cast h
+end
+
+lemma lt_add_one_iff_lt {x y : enat} (hx : x ≠ ⊤) : x < y + 1 ↔ x ≤ y :=
+begin
+  split, exact le_of_lt_add_one,
+  intro h, rcases ne_top_iff.mp hx with ⟨m, rfl⟩,
+  induction y using enat.cases_on with n, { rw [top_add], apply coe_lt_top },
+  apply_mod_cast nat.lt_succ_of_le, apply_mod_cast h
+end
+
+section mul
+
+instance : has_mul enat := ⟨λ x y, ⟨x.dom ∧ y.dom, λ h, get x h.1 * get y h.2⟩⟩
```

Comment:
Maybe it is good to explicitly give lemmas for `top_mul` and `mul_top`, so that users realize that it is absorbing (even for `0`).

Reply:
I deleted this instance. In the end I didn't use it, because the multiplicity of `a ^ k` was already in mathlib, and defined using `add_monoid.smul`, which has the convention that `0 • ⊤ = 0`. So this multiplication would just be confusing.